AP_InertialSensor: support logging of batches of raw IMU data

peterbarker opened the pull request:
Full-on logging of raw IMU data is too expensive for people to leave on all the time, in terms of CPU time, SD card and SD bandwidth.

This PR implements an alternative to the current raw-IMU logging.  Sensor data is taken at full rate into an allocated buffer (in the relevant sensor thread).  This data is then streamed out to the SD card in the main thread at a constant rate (1kB/s).  Each IMU sensor is sampled in turn.

The pymavlink tools directory contains `mavfft_ibs.py` (https://github.com/ArduPilot/pymavlink/blob/master/tools/mavfft_isb.py), a tool to extract the samples from a log and create a frequency analysis plot from that data.

A n Issue has been created for Mission Planner (https://github.com/ArduPilot/MissionPlanner/issues/1647).

tridge commented on libraries/AP_InertialSensor/BatchSampler.cpp:
```diff
@@ -0,0 +1,198 @@
+            if (instance >= _imu._accel_count) {
+                type = IMU_SENSOR_TYPE_GYRO;
+                instance = 0;
+                multiplier = multiplier_accel;
```
I think this should be multiplier_gyro

tridge commented on libraries/AP_InertialSensor/BatchSampler.cpp:
```diff
@@ -0,0 +1,198 @@
+            if (instance >= _imu._gyro_count) {
+                type = IMU_SENSOR_TYPE_ACCEL;
+                instance = 0;
+                multiplier = multiplier_gyro;
```
this should be multipler_accel
I think this should the code should be restructured a bit to make it clearer. Maybe a first switch to change the type, then a 2nd switch to setup the new multuplier

peterbarker replied:
Will fix.  I've just factored out the setting of multiplier.  It doesn't help much.

tridge commented on libraries/AP_InertialSensor/BatchSampler.cpp:
```diff
@@ -0,0 +1,198 @@
+        if (_sensor_mask & (1U<<instance)) {
+            break;
+        }
+    } while (instance != old_instance);
```
don't we risk looping forever here?

peterbarker replied:
Only if the number of gyros or accels change, I think.

I've changed this to a pair of for loops, and used the assumption (as discussed) that there will always be an equal number of gyros and accels.

tridge commented on libraries/AP_InertialSensor/BatchSampler.cpp:
```diff
@@ -0,0 +1,198 @@
+        return;
+    }
+    const uint64_t now = AP_HAL::micros64();
+    if (last_sent_ms - now < push_interval_ms) {
```
it is not safe to subtract uint64_t and uint32_t - both must be the same width

tridge replied:
also need to be the same units! don't compare milliseconds and microseconds

peterbarker replied:
Will fix.

peterbarker commented:
@OXINARF why the "NeedToChangeGCS" tag?  Certainly there could be benefits to changing them, but it should be a prereq for merging this.

tridge commented on libraries/AP_InertialSensor/AP_InertialSensor.h:
```diff
@@ -268,6 +273,74 @@ class AP_InertialSensor : AP_AccelCal_Client
     uint32_t get_last_update_usec(void) const { return _last_update_usec; }
 
+    enum IMU_SENSOR_TYPE {
+        IMU_SENSOR_TYPE_ACCEL,
```
minor point, given this forms part of the log data (so part of a protocol), I think giving the specific values with =0 and =1 is worthwhile in the enum

peterbarker replied:
Done.

peterbarker commented:
@tridge I've updated the expected time taken for the scheduled `ins.periodic()` based on `perf`'s output.

OXINARF commented:
> why the "NeedToChangeGCS" tag? Certainly there could be benefits to changing them, but it should be a prereq for merging this.

@peterbarker What happens when a GCS tries to analyse a log that has an *a* format in some message? It can certainly try to go over the message, but that's a fallback, not the correct way to operate.

peterbarker commented:
On Thu, 5 Oct 2017, Francisco Ferreira wrote:

>
>       why the "NeedToChangeGCS" tag? Certainly there could be benefits
>       to changing them, but it should be a prereq for merging this.
> 
> @peterbarker What happens when a GCS tries to analyse a log that has an a
> format in some message? It can certainly try to go over the message, but
> that's a fallback, not the correct way to operate.

Skipping over the message is the only practical way forward.  Thankfully 
we know the size of the message so that's easy enough.

We had this problem when we added the Q and q types - that was a lot more 
disruptive as we changes to using TimeUS with those :-)

OXINARF commented:
> Skipping over the message is the only practical way forward.  Thankfully
we know the size of the message so that's easy enough.

If a GCS is changing code to skip over these specific messages then it's much easier to support the *a* format instead.
Just to clarify: I'm not saying that the PR should be blocked until all GCS have adapted, just saying that GCS need to be changed.

peterbarker commented:
On Fri, 6 Oct 2017, Francisco Ferreira wrote:

>       Skipping over the message is the only practical way forward.
>       Thankfully
>       we know the size of the message so that's easy enough.
> 
> If a GCS is changing code to skip over these specific messages then it's
> much easier to support the a format instead.
> Just to clarify: I'm not saying that the PR should be blocked until all GCS
> have adapted, just saying that GCS need to be changed.

Now I'm with you.

You're almost certainly correct, at least in some cases.  We probably 
should have learnt when we added 'Q' and 'q' support to skip messages with 
fields we didn't understand, but I'm pretty sure the parsers I've worked 
with don't do that....

peterbarker commented:
On Fri, 6 Oct 2017, Francisco Ferreira wrote:

> > @@ -106,6 +108,17 @@ class DataFlash_Class
>      void Log_Write_RFND(const RangeFinder &rangefinder);
>      void Log_Write_IMU(const AP_InertialSensor &ins);
>      void Log_Write_IMUDT(const AP_InertialSensor &ins, uint64_t time_us, ui
> nt8_t imu_mask);
> +    bool Log_Write_ISBH(uint16_t seqno,
> +                        AP_InertialSensor::IMU_SENSOR_TYPE sensor_type,
> +                        uint8_t instance,
> +                        uint16_t multiplier,
> +                        uint64_t sample_us,
> +                        float sample_rate_hz);
> +    bool Log_Write_ISBD(uint16_t isb_seqno,
> +                        uint16_t seqno,
> +                        const int16_t x[32],
> +                        const int16_t y[32],
> +                        const int16_t z[32]);
> 
> const? 😜

I *think* that's all correct?  Not const'ing stack stuff but const'ing the 
more complicated ones....

> In libraries/AP_Math/AP_Math.h:
> 
> > @@ -154,7 +154,7 @@ inline int32_t constrain_int32(const int32_t amt, cons
> t int32_t low, const int32
>  }
>
>  // degrees -> radians
> -static inline float radians(float deg)
> +static inline constexpr float radians(float deg)
> 
> It would be nice to put the same in degrees

Will do.

> ____________________________________________________________________________
> 
> In libraries/AP_InertialSensor/BatchSampler.cpp:
> 
> > +
> +
> +extern const AP_HAL::HAL& hal;
> +void AP_InertialSensor::BatchSampler::init()
> +{
> +    if (_sensor_mask == 0) {
> +        return;
> +    }
> +    if (_required_count <= 0) {
> +        return;
> +    }
> +
> +    _required_count -= _required_count % 32; // round down to nearest multi
> ple of 32
> +
> +    const uint32_t total_allocation = 3*_required_count*sizeof(uint16_t);
> +    gcs().send_text(MAV_SEVERITY_WARNING, "INS: alloc %u bytes for ISB (fre
> e=%u)", total_allocation, hal.util->available_memory());
> 
> Does this really need to be a warning? Even info looks too much, looks more
> like debug.

Changed to debug.

> ____________________________________________________________________________
> 
> In libraries/AP_InertialSensor/BatchSampler.cpp:
> 
> > +
> +    if (type == IMU_SENSOR_TYPE_ACCEL) {
> +        // we have logged accelerometers, now log gyros:
> +        type = IMU_SENSOR_TYPE_GYRO;
> +        multiplier = multiplier_gyro;
> +        return;
> +    }
> +
> +    // log for accel sensor:
> +    type = IMU_SENSOR_TYPE_ACCEL;
> +    multiplier = multiplier_accel;
> +
> +    // move to next IMU backend:
> +
> +    // we assume the number of gyros and accels is the same:
> +    const uint8_t max_count = MIN(_imu._accel_count, _imu._gyro_count);
> 
> This seems strange:
>
>  *  the comment says we assume the number of gyros and accels is the same
>     but then we are taking a minimum between them (if we assume they are the
>     same, just take one of them, no need for the comparison)

The comment's correct, but obviously needs some help.  My first 
implementation didn't assume the counts were equal, but handling that 
wasn't pretty.  The min() just stops us dying if the counts are not equal.

I've adjusted the comment to indicate why we're doing the min() in 
particular.

>  *  the variable is named max_count but it is taking the minimum number of
>     accels/gyros

Heh.  Changed it to simply "_count"

Thanks!

peterbarker commented:
On Mon, 9 Oct 2017, Francisco Ferreira wrote:

> @OXINARF approved this pull request.
> 
> LGTM.
> 
> Do you have any plans to enable this by default in the future?

That's indeed the intent.  The IMU mask was primarily added at this point 
to give us a way of disabling the feature until we've got some "all good, 
no fires, floods or plagues of locusts were created when I enabled this 
option" reports.

It's possible that when we do enable it by default that we might exclude 
some known-bad-and-unusable IMUs based on board type.

tridge commented on libraries/AP_InertialSensor/BatchSampler.cpp:
```diff
@@ -0,0 +1,216 @@
+        return;
+    }
+    const uint32_t now = AP_HAL::millis();
+    if (last_sent_ms - now < push_interval_ms) {
```
this should be now - last_sent_ms

tridge commented on libraries/AP_InertialSensor/BatchSampler.cpp:
```diff
@@ -0,0 +1,216 @@
+        measurement_started_us = sample_us;
+    }
+
+    data_x[data_write_offset] = multiplier*_sample.x;
```
is there some reason to not use a Vector3i ?

tridge commented on libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp:
```diff
@@ -200,8 +200,17 @@ void AP_InertialSensor_Backend::_notify_new_gyro_raw_sample(uint8_t instance,
+        // should not have been called
+        return;
+    }
+    if (should_log_imu_raw()) {
```
can we remove the old log_GYRO and log_ACCEL now?

tridge commented:
I think this is ready to go in
